src/common/cli_wrapper.h

```diff
@@ -0,0 +1,275 @@
+#pragma once
+
+#include "3rd_party/CLI/CLI.hpp"
+#include "3rd_party/yaml-cpp/yaml.h"
+#include "common/any_type.h"
+#include "common/definitions.h"
+#include "common/logging.h"
+
+#include <iostream>
+#include <map>
+#include <string>
+
+namespace marian {
+namespace cli {
+
+// try to determine the width of the terminal
+static uint16_t guess_terminal_width(uint16_t max_width = 0,
+                                     uint16_t default_width = 180);
+
+namespace validators {
+const CLI::detail::ExistingFileValidator file_exists;
+const CLI::detail::ExistingDirectoryValidator dir_exists;
+const CLI::detail::ExistingPathValidator path_exists;
+
+const CLI::detail::NonexistentPathValidator path_not_exists;
+
+typedef CLI::Range range;
+}
+
+class CLIFormatter : public CLI::Formatter {
+public:
+  CLIFormatter(size_t columnWidth, size_t screenWidth);
+  virtual std::string make_option_desc(const CLI::Option *) const;
+
+private:
+  size_t screenWidth_{0};
+};
+
+class CLIWrapper {
+private:
+  // Stores option variables
+  std::map<std::string, Ptr<any_type>> vars_;
+  // Stores option objects
+  std::map<std::string, CLI::Option *> opts_;
+  // Command-line argument parser
+  Ptr<CLI::App> app_;
+  // Stores options as YAML object
+  YAML::Node config_;
+
+  // Name of the default option group
+  std::string defaultGroup_{""};
+  // Name of the current option group
+  std::string currentGroup_{""};
+
+  static std::string failureMessage(const CLI::App *app, const CLI::Error &e);
+
+  std::string keyName(const std::string &args) const {
+    return std::get<1>(CLI::detail::get_names(CLI::detail::split_names(args)))
+        .front();
+  }
+
+public:
+  /**
+   * @brief Create an instance of the command-line parser
```

Comment:
I would expect such a constructor to take argc and argv as arguments. Could you add a comment where those are fed in, and why this class must be stateful (where argc/argv are fed in later in a state update rather than creating an immutable class that gets constructed from the command line and is its representation)?

Maybe just roughly outline its use; something like First, call add() to create all teh options; then call ... to parse the cmd line, to make orientation for the reader easier.

Maybe the concept of option groups shoudl be touched upon here as well?

Reply:
Done